app/src/main/scala/org/alephium/explorer/persistence/queries/InputsQueries.scala

```diff
@@ -0,0 +1,165 @@
+
+package org.alephium.explorer.persistence.queries
+
+import scala.concurrent.ExecutionContext
+
+import com.typesafe.scalalogging.StrictLogging
+import slick.basic.DatabaseConfig
+import slick.dbio.DBIOAction
+import slick.jdbc.JdbcProfile
+
+import org.alephium.explorer.Hash
+import org.alephium.explorer.api.model._
+import org.alephium.explorer.persistence._
+import org.alephium.explorer.persistence.model._
+import org.alephium.explorer.persistence.schema._
+import org.alephium.util.U256
+
+trait InputsQueries extends InputSchema with OutputSchema with StrictLogging {
+
+  implicit def executionContext: ExecutionContext
+  val config: DatabaseConfig[JdbcProfile]
+  import config.profile.api._
+
+  private val mainInputs  = inputsTable.filter(_.mainChain)
+  private val mainOutputs = outputsTable.filter(_.mainChain)
+
+  def insertTxPerAddressFromInputs(inputs: Seq[InputEntity],
+                                   outputs: Seq[OutputEntity]): DBActionW[Seq[InputEntity]] = {
+
+    val inputsWithMaybeAddress = inputs.map(in => (in.address, in))
+
+    val assets = inputsWithMaybeAddress
+      .collect {
+        case (Some(address), input)
+            if !outputs.exists(out => out.address == address && out.txHash == input.txHash) =>
+          (address, input)
+      }
+      .distinctBy { case (address, input) => (address, input.txHash) }
+
+    val others = inputsWithMaybeAddress.collect {
+      case (None, input) => input
+    }
+
+    for {
+      _ <- insertInputWithAddress(assets)
+      inputsToUpdate <- DBIOAction
+        .sequence(others.map { input =>
+          insertTxPerAddressFromInput(input).map { i =>
+            if (i != 1) {
+              Seq(input)
+            } else {
+              Seq.empty
+            }
+          }
+        })
+        .map(_.flatten)
+    } yield inputsToUpdate
+  }
+
+  def insertTxPerAddressFromInput(input: InputEntity): DBActionW[Int] = {
+    sqlu"""
+      INSERT INTO transaction_per_addresses (hash, block_hash, timestamp, tx_index, address, main_chain)
+      (SELECT ${input.txHash}, ${input.blockHash}, ${input.timestamp}, ${input.txIndex}, address, main_chain FROM outputs WHERE key = ${input.outputRefKey})
+      ON CONFLICT (hash, block_hash, address) DO UPDATE
+      SET address = EXCLUDED.address
+    """
+  }
+
+  def insertInputWithAddress(inputs: Seq[(Address, InputEntity)]): DBActionW[Int] = {
+    if (inputs.nonEmpty) {
+      val values = inputs
+        .map {
+          case (address, input) =>
+            val instant = java.time.Instant.ofEpochMilli(input.timestamp.millis)
+            s"('\\x${input.txHash}', '\\x${input.blockHash}', '${instant}', ${input.txIndex}, '${address}', ${input.mainChain}) "
+        }
+        .mkString(",\n")
+
+      sqlu"""
+      INSERT INTO transaction_per_addresses (hash, block_hash, timestamp, tx_index, address, main_chain)
+      VALUES #$values
+      ON CONFLICT (hash, block_hash, address) DO NOTHING
```

Comment:
There should be no conflict. We should throw an error in this case

Reply:
yes we could as when we start syncing we step back few hours to be sure we don't miss a block, so we do re-insert.